app/xwd/man/xwd.man
.\" Copyright 1988, 1994, 1998  The Open Group
.\"
.\" Permission to use, copy, modify, distribute, and sell this software and its
.\" documentation for any purpose is hereby granted without fee, provided that
.\" the above copyright notice appear in all copies and that both that
.\" copyright notice and this permission notice appear in supporting
.\" documentation.
.\"
.\" The above copyright notice and this permission notice shall be included
.\" in all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR
.\" OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
.\" ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
.\" OTHER DEALINGS IN THE SOFTWARE.
.\"
.\" Except as contained in this notice, the name of The Open Group shall
.\" not be used in advertising or otherwise to promote the sale, use or
.\" other dealings in this Software without prior written authorization
.\" from The Open Group.
.\"
.TH XWD 1 2025-04-12 __xorgversion__
.SH NAME
xwd - dump an image of an X window
.SH SYNOPSIS
.nf
\fBxwd\fP [\fB\-debug\fP] [\fB\-help\fP] [\fB\-nobdrs\fP] [\fB\-out\fP \fIfile\fP] [\fB\-xy\fP] [\fB\-frame\fP] [\fB\-add\fP \fIvalue\fP]
    [\fB\-root\fP | \fB\-id\fP \fIwdid\fP | \fB\-name\fP \fIwdname\fP] [\fB\-icmap\fP] [\fB\-screen\fP] [\fB\-silent\fP]
    [\fB\-display\fP \fIhost:dpy\fP] [\fB\-version\fP]
.fi
.SH DESCRIPTION
.B Xwd
is an X Window System window dumping utility.
.B Xwd
allows X users to store window images in a specially formatted dump file.
This file can then be read by various other X utilities for
redisplay, printing, editing, formatting, archiving, image processing, etc.
The target window is selected by clicking the pointer in the desired window.
The keyboard bell is rung once at the beginning of the dump and twice when
the dump is completed.
.SH OPTIONS
.TP 8
.B "\-debug"
Enable debug mode.
.TP 8
.B "\-d, \-display \fIhost:dpy\fP"
This argument allows you to specify the server to connect to; see
.BR X (__miscmansuffix__).
.TP 8
.B "\-help"
Print out the \[lq]Usage:\[rq] command syntax summary.
.TP 8
.B "\-nobdrs"
This argument specifies that the window dump should not include the
pixels that compose the X window border.
This is useful in situations where you may wish to include the
window contents in a document as an illustration.
.TP 8
.B "\-out \fIfile\fP"
This argument allows the user to explicitly specify the output
file on the command line.
The default is to output to standard out.
.TP 8
.B "\-xy"
This option applies to color displays only.
It selects \[lq]XY\[rq] format dumping instead of
the default \[lq]Z\[rq] format.
.TP 8
.B "\-add \fIvalue\fP"
This option specifies a signed value to be added to every pixel.
.TP 8
.B "\-frame"
This option indicates that the window manager frame should be included when
manually selecting a window.
.TP 8
.B "\-root"
This option indicates that the root window
should be selected for the window dump,
without requiring the user to select a window with the pointer.
.TP 8
.B "\-id \fIwdid\fP"
This option indicates that the window with the specified resource id
should be selected for the window dump,
without requiring the user to select a window with the pointer.
.TP 8
.B "\-name \fIwdname\fP"
This option indicates that the window with the specified WM_NAME property
should be selected for the window dump,
without requiring the user to select a window with the pointer.
.TP 8
.B "\-icmap"
Normally the colormap of the chosen window is used to obtain RGB values.
This option forces the first installed colormap of the screen to be used
instead.
.TP 8
.B "\-screen"
This option indicates that the GetImage request used to obtain the image
should be done on the root window,
rather than directly on the specified window.
In this way, you can obtain pieces of other windows that overlap
the specified window, and more importantly, you can capture menus or other
popups that are independent windows but appear over the specified window.
.TP 8
.B "\-silent"
Operate silently, i.e., don't ring any bells before and after dumping
the window.
.TP 8
.B "\-version"
This option indicates that
.B xwd
should print its version information and exit.
.SH ENVIRONMENT
.TP 8
.B DISPLAY
To get default host and display number.
.SH FILES
.TP 8
.B XWDFile.h
X Window Dump File format definition file.
.SH SEE ALSO
.BR xpr (__appmansuffix__),
.BR xwud (__appmansuffix__),
.BR X (__miscmansuffix__)
.SH AUTHORS
Tony Della Fera, Digital Equipment Corp., MIT Project Athena
.br
William F. Wyatt, Smithsonian Astrophysical Observatory


xserver/ChangeLog
commit e597f0119cd69b6d9edf86d06d941468f90d8e6d
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 27 21:07:17 2012 -0800

    configure.ac: 1.11.4
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit 6c491463ee7a3637c8aae7ddae9625b36b792078
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 20 23:54:58 2012 -0800

    1.11.3.902 (1.11.4 RC2)
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit b747a91485184649c05aba6fa10734f70e926032
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Wed Jan 18 11:47:21 2012 -0800

    XQuartz: Bump bundle version to 2.7.1
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit a1ed67fa0551ee2b39d002d3b786a2017b37c7ac
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Fri Jan 6 13:20:45 2012 +1000

    render: don't bother with animated cursors on floating slaves (#39989)
    
    X.Org Bug 39989 <http://bugs.freedesktop.org/show_bug.cgi?id=39989>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit bbb6b8c834e0e1491ca14403b5d0840dd14380d3)

commit f89c39e6c9e20ffc3c1ae228ee601c2ab9007ff4
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Mon Jan 9 15:04:10 2012 +1000

    dix: Update pointer limits for floating devices too (#43635)
    
    When the screen is restructured, the pointer limits need to be reset for
    floating slave devices as well, not just for master pointers. Only skip
    devices that don't have a cursor (attached slaves and keyboard)
    
    Bug reproducer: float an absolute slave device, rotate the screen - the
    device is now confined to a section of the screen only.
    
    X.Org Bug 43635 <http://bugs.freedesktop.org/show_bug.cgi?id=43635>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Chase Douglas <chase.douglas@canonical.com>
    (cherry picked from commit 35bd77e9d0701daae87d681900d749604fc6471f)

commit dea9377a257a402b2c7fe4a96c958d951a7d4579
Author: Zhigang Gong <zhigang.gong@linux.intel.com>
Date:   Wed Jan 4 07:01:19 2012 +0000

    mi/mibitblt: Fix an overflow bug of bit shift.
    
    When depth equal to 32 and planeMask equal to 0, the overflow will
    occur and cause the pixmap can't be cleared. There are some test
    cases in XTS hit this bug, and this fix can eliminate the corresponding
    failures.
    
    Signed-off-by: Zhigang Gong <zhigang.gong@linux.intel.com>
    Reviewed-by: Adam Jackson <ajax@redhat.com>
    Signed-off-by: Keith Packard <keithp@keithp.com>
    (cherry picked from commit f4956faab9ccf9aba6cf9603f4489f5dad19a347)

commit d2db5100b1bafeeea8c1febd390631c8f8abb094
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Fri Oct 28 21:29:50 2011 -0700

    Convert DetermineClientCmd to use strdup instead of malloc+strncpy
    
    *cmdname is initialized to NULL earlier in the function, so it's
    okay to overwrite it with NULL if strdup fails, don't need that
    extra check.
    
    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit 780133f9ae7fada462714b47e79d26075bbd9abe)

commit 5b28bcb340e279e7f4e76617e86b439092367690
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:23:12 2011 -0500

    dix: Tune dixLookupDrawable for success
    
    The vast vast vast majority of resource lookups are successful.  Move some
    work to the error paths so we don't punish success.
    
    Before:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1148346.9/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2091666.1/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 8f9bdfd293ad8e45755efe8d764b4dcc2a724f51)

commit e1ea93fec93b13d18df4451210aaaf322b769b6a
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 20:23:40 2011 -0500

    dix: Tune dtrace hooks around Dispatch
    
    Don't call LookupMajorName if the hooks aren't active, it's quite expensive.
    
    Before:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2012238.6/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 83a98543b58c661a22574a6f8d8f9d777c0955b8)

commit 9eaa0b5c89f645c50ea90e07e8b49ebacf4349a0
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:00:05 2011 -0500

    xace: ricer tuning for XaceHook
    
    gcc doesn't want to hoist the check for XaceHooks[hook] != NULL above the
    varargs code for some reason, so do it ourselves.
    
    Before:
    40000000 trep @   0.0010 msec (1050420.2/sec): PutImage 10x10 square

app/xwd/xwd.c
/*

Copyright 1987, 1998  The Open Group

Permission to use, copy, modify, distribute, and sell this software and its
documentation for any purpose is hereby granted without fee, provided that
the above copyright notice appear in all copies and that both that
copyright notice and this permission notice appear in supporting
documentation.

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of The Open Group shall not be
used in advertising or otherwise to promote the sale, use or other dealings
in this Software without prior written authorization from The Open Group.

*/

/*
 * xwd.c MIT Project Athena, X Window system window raster image dumper.
 *
 * This program will dump a raster image of the contents of a window into a
 * file for output on graphics printers or for other uses.
 *
 *  Author:	Tony Della Fera, DEC
 *		17-Jun-85
 *
 *  Modification history:
 *
 *  11/14/86 Bill Wyatt, Smithsonian Astrophysical Observatory
 *    - Removed Z format option, changing it to an XY option. Monochrome
 *      windows will always dump in XY format. Color windows will dump
 *      in Z format by default, but can be dumped in XY format with the
 *      -xy option.
 *
 *  11/18/86 Bill Wyatt
 *    - VERSION 6 is same as version 5 for monchrome. For colors, the
 *      appropriate number of Color structs are dumped after the header,
 *      which has the number of colors (=0 for monochrome) in place of the
 *      V5 padding at the end. Up to 16-bit displays are supported. I
 *      don't yet know how 24- to 32-bit displays will be handled under
 *      the Version 11 protocol.
 *
 *  6/15/87 David Krikorian, MIT Project Athena
 *    - VERSION 7 runs under the X Version 11 servers, while the previous
 *      versions of xwd were are for X Version 10.  This version is based
 *      on xwd version 6, and should eventually have the same color
 *      abilities. (Xwd V7 has yet to be tested on a color machine, so
 *      all color-related code is commented out until color support
 *      becomes practical.)
 */

/*%
 *%    This is the format for commenting out color-related code until
 *%  color can be supported.
%*/

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <stdio.h>
#include <errno.h>
#include <X11/Xos.h>
#include <stdlib.h>

#include <X11/Xlib.h>
#include <X11/Xutil.h>

#include "X11/XWDFile.h"

#define FEEP_VOLUME 0

/* Include routines to do parsing */
#include "dsimple.h"
#include "list.h"
#include "wsutils.h"
#include "multiVis.h"

#ifdef XKB
#include <X11/extensions/XKBbells.h>
#endif

/* Setable Options */

static int format = ZPixmap;
static Bool nobdrs = False;
static Bool on_root = False;
static Bool standard_out = True;
static Bool debug = False;
static Bool silent = False;
static Bool use_installed = False;
static long add_pixel_value = 0;


extern int main(int, char **);
extern void Window_Dump(Window, FILE *);
extern int Image_Size(XImage *);
extern int Get_XColors(XWindowAttributes *, XColor **);
extern void _swapshort(register char *, register unsigned);
extern void _swaplong(register char *, register unsigned);
static long parse_long(char *);
static int Get24bitDirectColors(XColor **);
static int ReadColors(Visual *, Colormap, XColor **);

static long
parse_long(char *s)
{
    long retval = 0L;
    int thesign = 1;

    if (s && s[0]) {
        switch (s[0]) {
        case '-':
            (void) sscanf(s + 1, "%lu", &retval);
            thesign = -1;
            break;
        case '0':
            (void) sscanf(s + 1, "%lo", &retval);
            break;
        case 'x':
        case 'X':
            (void) sscanf(s + 1, "%lx", &retval);
            break;
        default:
            (void) sscanf(s, "%lu", &retval);
            break;
        }
    }
    return (thesign * retval);
}

int
main(int argc, char **argv)
{
    register int i;
    Window target_win;
    FILE *out_file = stdout;
    Bool frame_only = False;

    INIT_NAME;

    /* Handle args that don't require opening a display */
    for (int n = 1; n < argc; n++) {
	const char *argn = argv[n];
	/* accept single or double dash for -help & -version */
	if (argn[0] == '-' && argn[1] == '-') {
	    argn++;
	}
	if (strcmp (argn, "-help") == 0) {
	    usage(NULL, EXIT_SUCCESS);
	    exit(0);
	}
	if (strcmp (argn, "-version") == 0) {
	    puts(PACKAGE_STRING);
	    exit(EXIT_SUCCESS);
	}
    }

    Setup_Display_And_Screen(&argc, argv);

    /* Get window select on command line, if any */
    target_win = Select_Window_Args(&argc, argv);

    for (i = 1; i < argc; i++) {
        if (!strcmp(argv[i], "-nobdrs")) {
            nobdrs = True;
            continue;
        }
        if (!strcmp(argv[i], "-debug")) {
            debug = True;
            continue;
        }
        if (!strcmp(argv[i], "-out")) {
            if (++i >= argc)
                usage("-out requires an argument", EXIT_FAILURE);
            if (!(out_file = fopen(argv[i], "wb")))
                Fatal_Error("Can't open output file as specified.");
            standard_out = False;
            continue;
        }
        if (!strcmp(argv[i], "-xy")) {
            format = XYPixmap;
            continue;
        }
        if (!strcmp(argv[i], "-screen")) {
            on_root = True;
            continue;
        }
        if (!strcmp(argv[i], "-icmap")) {
            use_installed = True;
            continue;
        }
        if (!strcmp(argv[i], "-add")) {
            if (++i >= argc)
                usage("-add requires an argument", EXIT_FAILURE);
            add_pixel_value = parse_long(argv[i]);
            continue;
        }
        if (!strcmp(argv[i], "-frame")) {
            frame_only = True;
            continue;
        }
        if (!strcmp(argv[i], "-silent")) {
            silent = True;
            continue;
        }
        fprintf(stderr, "%s: unrecognized argument '%s'\n",
                program_name, argv[i]);
        usage(NULL, EXIT_FAILURE);
    }
#ifdef WIN32
    if (standard_out)
        _setmode(fileno(out_file), _O_BINARY);
#endif

    /*
     * Let the user select the target window.
     */
    if (target_win == None)
        target_win = Select_Window(dpy, !frame_only);

    /*
     * Dump it!
     */
    Window_Dump(target_win, out_file);

    XCloseDisplay(dpy);
    if (fclose(out_file)) {
        perror("xwd");
        exit(1);
    }
    exit(0);
}

static int
Get24bitDirectColors(XColor **colors)
{
    int i, ncolors = 256;
    XColor *tcol;

    *colors = tcol = malloc(sizeof(XColor) * ncolors);
    if (tcol == NULL) {
        perror("xwd");
        exit(1);
    }

    for (i = 0; i < ncolors; i++) {
        tcol[i].pixel = i << 16 | i << 8 | i;
        tcol[i].red = tcol[i].green = tcol[i].blue = i << 8 | i;
        tcol[i].flags = 0;
    }

    return ncolors;
}

/*
 * Window_Dump: dump a window to a file which must already be open for
 *              writing.
 */

void
Window_Dump(Window window, FILE *out)
{
    unsigned long swaptest = 1;
    XColor *colors;
    unsigned buffer_size;
    size_t win_name_size;
    CARD32 header_size;
    int ncolors, i;
    char *win_name;
    char default_win_name[] = "xwdump";
    Bool got_win_name;
    XWindowAttributes win_info;
    XImage *image;
    int absx, absy, x, y;
    unsigned width, height;
    int dwidth, dheight;
    Window dummywin;
    XWDFileHeader header;
    XWDColor xwdcolor;

    int transparentOverlays, multiVis;
    int numVisuals;
    XVisualInfo *pVisuals;
    int numOverlayVisuals;
    OverlayInfo *pOverlayVisuals;
    int numImageVisuals;
    XVisualInfo **pImageVisuals;
    list_ptr vis_regions;       /* list of regions to read from */
    list_ptr vis_image_regions;
    Visual vis_h, *vis;
    int allImage = 0;

    /*
     * Inform the user not to alter the screen.
     */
    if (!silent) {
#ifdef XKB
        XkbStdBell(dpy, None, 50, XkbBI_Wait);
#else
        XBell(dpy, FEEP_VOLUME);
#endif
        XFlush(dpy);
    }

    /*
     * Get the parameters of the window being dumped.
     */
    if (debug)
        outl("xwd: Getting target window information.\n");
    if (!XGetWindowAttributes(dpy, window, &win_info))
        Fatal_Error("Can't get target window attributes.");

    /* handle any frame window */
    if (!XTranslateCoordinates(dpy, window, RootWindow(dpy, screen), 0, 0,
                               &absx, &absy, &dummywin)) {
        fprintf(stderr,
                "%s:  unable to translate window coordinates (%d,%d)\n",
                program_name, absx, absy);
        exit(1);
    }
    win_info.x = absx;
    win_info.y = absy;
    width = win_info.width;
    height = win_info.height;

    if (!nobdrs) {
        absx -= win_info.border_width;
        absy -= win_info.border_width;
        width += (2 * win_info.border_width);
        height += (2 * win_info.border_width);
    }
    dwidth = DisplayWidth(dpy, screen);
    dheight = DisplayHeight(dpy, screen);

    /* clip to window */
    if (absx < 0) {
        width += absx;
        absx = 0;
    }
    if (absy < 0) {
        height += absy;
        absy = 0;
    }
    if (absx + width > dwidth)
        width = dwidth - absx;
    if (absy + height > dheight)
        height = dheight - absy;

    if (!XFetchName(dpy, window, &win_name) || !win_name || !win_name[0]) {
        win_name = default_win_name;
        got_win_name = False;
    }
    else {
        got_win_name = True;
    }

    /* sizeof(char) is included for the null string terminator. */
    win_name_size = strlen(win_name) + sizeof(char);

    /*
     * Snarf the pixmap with XGetImage.
     */

    x = absx - win_info.x;
    y = absy - win_info.y;

    multiVis = GetMultiVisualRegions(dpy, RootWindow(dpy, screen),
                                     absx, absy,
                                     width, height, &transparentOverlays,
                                     &numVisuals, &pVisuals,
                                     &numOverlayVisuals, &pOverlayVisuals,
                                     &numImageVisuals, &pImageVisuals,
                                     &vis_regions, &vis_image_regions,
                                     &allImage);
    if (on_root || multiVis) {
        if (!multiVis)
            image = XGetImage(dpy, RootWindow(dpy, screen), absx, absy,
                              width, height, AllPlanes, format);
        else
            image = ReadAreaToImage(dpy, RootWindow(dpy, screen), absx, absy,
                                    width, height,
                                    numVisuals, pVisuals,
                                    numOverlayVisuals, pOverlayVisuals,
                                    numImageVisuals, pImageVisuals,
                                    vis_regions, vis_image_regions,
                                    format, allImage);
    }
    else
        image = XGetImage(dpy, window, x, y, width, height, AllPlanes, format);
    if (!image) {
        fprintf(stderr, "%s:  unable to get image at %dx%d+%d+%d\n",
                program_name, width, height, x, y);
        exit(1);
    }

    if (add_pixel_value != 0)
        XAddPixel(image, add_pixel_value);

    /*
     * Determine the pixmap size.
     */
    buffer_size = Image_Size(image);

    if (debug)
        outl("xwd: Getting Colors.\n");

    if (!multiVis) {
        ncolors = Get_XColors(&win_info, &colors);
        vis = win_info.visual;
    }
    else {
        ncolors = Get24bitDirectColors(&colors);
        initFakeVisual(&vis_h);
        vis = &vis_h;
    }
    /*
     * Inform the user that the image has been retrieved.
     */
    if (!silent) {
#ifdef XKB
        XkbStdBell(dpy, window, FEEP_VOLUME, XkbBI_Proceed);
        XkbStdBell(dpy, window, FEEP_VOLUME, XkbBI_RepeatingLastBell);
#else
        XBell(dpy, FEEP_VOLUME);
        XBell(dpy, FEEP_VOLUME);
#endif
        XFlush(dpy);
    }

    /*
     * Calculate header size.
     */
    if (debug)
        outl("xwd: Calculating header size.\n");
    header_size = SIZEOF(XWDheader) + (CARD32) win_name_size;

    /*
     * Write out header information.
     */
    if (debug)
        outl("xwd: Constructing and dumping file header.\n");
    memset(&header, 0, SIZEOF(XWDheader));
    header.header_size = (CARD32) header_size;
    header.file_version = (CARD32) XWD_FILE_VERSION;
    header.pixmap_format = (CARD32) format;
    header.pixmap_depth = (CARD32) image->depth;
    header.pixmap_width = (CARD32) image->width;
    header.pixmap_height = (CARD32) image->height;
    header.xoffset = (CARD32) image->xoffset;
    header.byte_order = (CARD32) image->byte_order;
    header.bitmap_unit = (CARD32) image->bitmap_unit;
    header.bitmap_bit_order = (CARD32) image->bitmap_bit_order;
    header.bitmap_pad = (CARD32) image->bitmap_pad;
    header.bits_per_pixel = (CARD32) image->bits_per_pixel;
    header.bytes_per_line = (CARD32) image->bytes_per_line;
    /****
    header.visual_class = (CARD32) win_info.visual->class;
    header.red_mask = (CARD32) win_info.visual->red_mask;
    header.green_mask = (CARD32) win_info.visual->green_mask;
    header.blue_mask = (CARD32) win_info.visual->blue_mask;
    header.bits_per_rgb = (CARD32) win_info.visual->bits_per_rgb;
    header.colormap_entries = (CARD32) win_info.visual->map_entries;
    *****/
    header.visual_class = (CARD32) vis->class;
    header.red_mask = (CARD32) vis->red_mask;
    header.green_mask = (CARD32) vis->green_mask;
    header.blue_mask = (CARD32) vis->blue_mask;
    header.bits_per_rgb = (CARD32) vis->bits_per_rgb;
    header.colormap_entries = (CARD32) vis->map_entries;

    header.ncolors = ncolors;
    header.window_width = (CARD32) win_info.width;
    header.window_height = (CARD32) win_info.height;
    header.window_x = absx;
    header.window_y = absy;
    header.window_bdrwidth = (CARD32) win_info.border_width;

    if (*(char *) &swaptest) {
        _swaplong((char *) &header, sizeof(header));
        for (i = 0; i < ncolors; i++) {
            _swaplong((char *) &colors[i].pixel, sizeof(CARD32));
            _swapshort((char *) &colors[i].red, 3 * sizeof(short));
        }
    }

    if (fwrite((char *) &header, SIZEOF(XWDheader), 1, out) != 1 ||
        fwrite(win_name, win_name_size, 1, out) != 1) {
        perror("xwd");
        exit(1);
    }

    /*
     * Write out the color maps, if any
     */

    if (debug)
        outl("xwd: Dumping %d colors.\n", ncolors);
    for (i = 0; i < ncolors; i++) {
        xwdcolor.pixel = colors[i].pixel;
        xwdcolor.red = colors[i].red;
        xwdcolor.green = colors[i].green;
        xwdcolor.blue = colors[i].blue;
        xwdcolor.flags = colors[i].flags;
        if (fwrite((char *) &xwdcolor, SIZEOF(XWDColor), 1, out) != 1) {
            perror("xwd");
            exit(1);
        }
    }

    /*
     * Write out the buffer.
     */
    if (debug)
        outl("xwd: Dumping pixmap.  bufsize=%d\n", buffer_size);

    /*
     *  This copying of the bit stream (data) to a file is to be replaced
     *  by an Xlib call which hasn't been written yet.  It is not clear
     *  what other functions of xwd will be taken over by this (as yet)
     *  non-existent X function.
     */
    if (fwrite(image->data, (int) buffer_size, 1, out) != 1) {
        perror("xwd");
        exit(1);
    }

    /*
     * free the color buffer.
     */

    if (debug && ncolors > 0)
        outl("xwd: Freeing colors.\n");
    if (ncolors > 0)
        free(colors);

    /*
     * Free window name string.
     */
    if (debug)
        outl("xwd: Freeing window name string.\n");
    if (got_win_name)
        XFree(win_name);

    /*
     * Free image
     */
    XDestroyImage(image);
}

/*
 * Report the syntax for calling xwd.
 */
void
usage(const char *errmsg, int exitval)
{
    if (errmsg != NULL)
        fprintf(stderr, "%s: %s\n", program_name, errmsg);

    fprintf(stderr,
            "Usage: %s [options] [-root| -id <wdid>| -name <wdname>] > mywddump\n",
            program_name);
    fprintf(stderr,
            "       %s [options] [-root| -id <wdid>| -name <wdname>] -out mywddump\n",
            program_name);
    fputs("Options:\n" "  -help                    Print this message\n"
          "  -version                 Print the program version and exit\n"
          "  -debug                   Enable debug mode\n"
          "  -d, -display <host:dpy>  Specify server to connect\n"
          "  -nobdrs                  Exclude window borders\n"
          "  -out <file>              Specify an output file\n"
          "  -xy                      Select XY dumping format for color displays\n"
          "  -add <value>             Add a signed value to every pixel\n"
          "  -frame                   Include window manager frame\n"
          "  -root                    Select the root window\n"
          "  -id <wdid>               Select a window by its resource id\n"
          "  -name <wdname>           Select a window by its WM_NAME property\n"
          "  -icmap                   Use the first colormap of the screen\n"
          "  -screen                  Send the request against the root window\n"
          "  -silent                  Don't ring any bells\n", stderr);
    exit(exitval);
}

/*
 * Determine the pixmap size.
 */

int
Image_Size(XImage *image)
{
    if (image->format != ZPixmap)
        return (image->bytes_per_line * image->height * image->depth);

    return (image->bytes_per_line * image->height);
}

#define lowbit(x) ((x) & (~(x) + 1))

static int
ReadColors(Visual *vis, Colormap cmap, XColor **colors)
{
    int i, ncolors;

    ncolors = vis->map_entries;

    if (!(*colors = malloc(sizeof(XColor) * ncolors)))
        Fatal_Error("Out of memory!");

    if (vis->class == DirectColor || vis->class == TrueColor) {
        Pixel red, green, blue, red1, green1, blue1;

        red = green = blue = 0;
        red1 = lowbit(vis->red_mask);
        green1 = lowbit(vis->green_mask);
        blue1 = lowbit(vis->blue_mask);
        for (i = 0; i < ncolors; i++) {
            (*colors)[i].pixel = red | green | blue;
            (*colors)[i].pad = 0;
            red += red1;
            if (red > vis->red_mask)
                red = 0;
            green += green1;
            if (green > vis->green_mask)
                green = 0;
            blue += blue1;
            if (blue > vis->blue_mask)
                blue = 0;
        }
    }
    else {
        for (i = 0; i < ncolors; i++) {
            (*colors)[i].pixel = i;
            (*colors)[i].pad = 0;
        }
    }

    XQueryColors(dpy, cmap, *colors, ncolors);

    return (ncolors);
}

/*
 * Get the XColors of all pixels in image - returns # of colors
 */
int
Get_XColors(XWindowAttributes *win_info, XColor **colors)
{
    int i, ncolors;
    Colormap cmap = win_info->colormap;

    if (use_installed)
        /* assume the visual will be OK ... */
        cmap = XListInstalledColormaps(dpy, win_info->root, &i)[0];
    if (!cmap)
        return (0);
    ncolors = ReadColors(win_info->visual, cmap, colors);
    return ncolors;
}

void
_swapshort(register char *bp, register unsigned n)
{
    char *ep = bp + n;

    while (bp < ep) {
        char c = *bp;
        *bp = *(bp + 1);
        bp++;
        *bp++ = c;
    }
}

void
_swaplong(register char *bp, register unsigned n)
{
    char *ep = bp + n;

    while (bp < ep) {
        char c = bp[3];
        bp[3] = bp[0];
        bp[0] = c;
        c = bp[2];
        bp[2] = bp[1];
        bp[1] = c;
        bp += 4;
    }
}
